manager/integration/tests/common.py

```diff
@@ -1950,10 +1952,19 @@ def crash_replica_processes(client, api, volname, replicas=None,
                        "' | grep -v grep | awk '{print $2}'`"
         exec_instance_manager(api, r.instanceManagerName, kill_command)
 
-        # FIXME: this is a work around, should check this in-parallel.
-        #        https://github.com/longhorn/longhorn/issues/4045
+        # XXX: check all replica status in-parallel but it might not make sure
+        #      that replica is failed after checking and quickly recovering.
+        #      https://github.com/longhorn/longhorn/issues/4045
         if wait_to_fail is True:
-            wait_for_replica_failed(client, volname, r['name'])
+            check_t = threading.Thread(target=wait_for_replica_failed,
```

Comment:
The assertion except can be caught in the main thread? how to know if `wait_for_replica_failed` failed?

Reply:
If can't, probably you cound consider using https://pypi.org/project/atomic/ or similar, create an atomic count and an ad-hoc closure to the `wait_for_replica_failed`, then increase the count if any failure.  (of course, you should modify `wait_for_replica_failed`.

When all threads are finished, check the count to understand if all replicas crashed already. 

Also, I don't think you should customize rety_* variables in `wait_for_replica_failed`. You can adjust it in the future if needed.

Reply:
It should be caught in the main thread, let me do some tests.

Reply:
I added a Thread class and the assertion except would be caught by this thread class then asserted the exception again.